fix(orc8r): Read from subscriber storage by IMSI

alexzurbonsen opened the pull request:
⚠️ Merge after #13212

## Summary

In this PR we enable NMS to read `subscriber_state` from the new `gateway_subscriber_states` table (#12949). It follows up on #13041 and implements step 4 from [#8621 - comment](https://github.com/magma/magma/issues/8621#issuecomment-1144762649).

We adapt the obsidian handlers of subscriberdb to read all states of type `lte.SubscriberStateType = 'subscriber_state'`
from the `gateway_subscriber_states` table instead of the `states` table. To that end we add a new method to the SubscriberStorage interface, which enables to read state by network ID and imsi (`GetSubscribersForIMSIs`). Accordingly we also add an index for network ID and imsi to the `gateway_subscriber_states` table.

The handlers that are getting `subscriber_state`, are now hard coded to always get it the new interface method by calling the `getSubscribersForIMSIs` function.

Tests are adapted accordingly.

## Test Plan

- [x] Check in postgres database that index was created.
- [x] run handlers_test.go
- [x] Test manually in NMS whether all relevant routes are loading state correctly
         - /magma/v1/lte/:network_id/subscribers (called in subscribers tab)
         - /magma/v1/lte/:network_id/subscribers/:subscriber_id (view of single subscriber)
         - /magma/v1/lte/:network_id/subscriber_state (called in subscribers tab)
         - /magma/v1/lte/:network_id/subscriber_state/:subscriber_id (called if autorefresh in single subscriber view is activated)
- [x] e2e test with simple integration test


## Additional Information

- Done in pairing with @sebathomas.

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

alexzurbonsen committed:
fix(orc8r): Add Get method that retrieves subscriber states by a list of IMSIs

Interface SubscriberStorage is extended to include the method.
Method is implemented to return subscriber states for all IMSIs
for a network if empty list is passed.

In case an IMSI is registered multiple times, only the most
recently registered entry is returned.

An index for network ID and IMSI is added to subscriberStorage to
make database requests fast.

Signed-off-by: Alexander zur Bonsen <alexander.zur.bonsen@tngtech.com>

alexzurbonsen commented on lte/cloud/go/services/subscriberdb/storage/subscriber_state_test.go:
```diff
@@ -101,4 +109,90 @@ func TestSubscriberStorage(t *testing.T) {
+}
+
+func (ss *subscriberStorage) setTestGatewaySubscriberStateWithIncreasingTimestamp(networkID string, gatewayID string, subscriberStates *GatewaySubscriberState) error {
+	timeSec := clock.Now()
```
Currently this method is only used to set the states in the final case of the test, the duplicate IMSIs. It uses system time (as `SetAllSubscribersForGateway`) and adds one second for every entry.
I.e. the first entry of `double_IMSI` is registered at system time (via `SetAllSubscribersForGateway`), the duplicate is registered at system time + 2 Seconds (via this method, `setTestGatewaySubscriberStateWithIncreasingTimestamp`).

More work, but may be less confusing could be to fix the time and set all states via this method? What do you think?

sebathomas replied:
Yes, I think it would be clearer to just give them explicit times and not rely on the system time + x seconds, and to create both states by the same method.

alexzurbonsen commented on lte/cloud/go/services/subscriberdb/storage/subscriber_state.go:
```diff
@@ -182,6 +195,57 @@ func (ss *subscriberStorage) DeleteSubscribersForGateway(networkID string, gatew
+			if err != nil {
+				return nil, fmt.Errorf("GetSubscribersForIMSIs, error unmarshaling state: %w", err)
+			}
+			if lastUpdated > lastUpdates[imsi] {
```
What do we do in the = case? The compared times are given in seconds. I am not sure whether it is realistic to have an IMSI registering with two gateways within one second. But maybe that is an edge case we should take care of?

sebathomas replied:
In the current implementation, we will randomly pick the state that is read first. I think that is fine. The conflict will only be temporary and should be solved when we get an update from one of the AGWs.

sebathomas commented on lte/cloud/go/services/subscriberdb/storage/subscriber_state_test.go:
```diff
@@ -101,4 +107,90 @@ func TestSubscriberStorage(t *testing.T) {
+		// the order of the imsis in imsisGw2 is important. The target
+		// doubled imsi "IMSI001010000000123", has to be the second entry,
+		// since its two entries are distinguished by the value of
+		// "active_duration_sec", which wouldn't differ otherwise
```
Maybe you can do something else to ensure they are different (e.g. modify the states after creation), and then assert on the entire state instead of only comparing `active_duration_sec`.

alexzurbonsen replied:
Sounds reasonable.

sebathomas commented on lte/cloud/go/services/subscriberdb/storage/subscriber_state_test.go:
```diff
@@ -101,4 +107,86 @@ func TestSubscriberStorage(t *testing.T) {
+
+func (ss *subscriberStorage) setTestGatewaySubscriberStateWithIncreasingTimestamp(t *testing.T, networkID string, gatewayID string, subscriberStates *GatewaySubscriberState, frozenTime int64) error {
+	clock.SetAndFreezeClock(t, time.Unix(frozenTime, 0))
+	defer clock.UnfreezeClock(t)
```
Can't you just put the ints into the database without setting or reading the clock?

alexzurbonsen replied:
Oh, right. Scusi. Didn't think of that.

alexzurbonsen committed:
Modify handlers that fetch states of type "subscriber_state"
to read from "gateway_subscriber_states" table.

Signed-off-by: Alexander zur Bonsen <alexander.zur.bonsen@tngtech.com>

alexzurbonsen committed:
Adapt handler tests send gateway_subscriber_state to gateway_subscriber_states
table and read from there as well.

Add respective annotation to subscriberdb test service to enable reindexing.

Signed-off-by: Alexander zur Bonsen <alexander.zur.bonsen@tngtech.com>

sebathomas commented on lte/cloud/go/services/subscriberdb/storage/subscriber_state_test.go:
```diff
@@ -101,4 +105,82 @@ func TestSubscriberStorage(t *testing.T) {
 	})
+
+	t.Run("Test getting subscriber states by imsi and getting all subscribers", func(t *testing.T) {
+
```
Nitpick: unneeded empty line

sebathomas commented on lte/cloud/go/services/subscriberdb/storage/subscriber_state_test.go:
```diff
@@ -14,10 +14,14 @@
+	"github.com/Masterminds/squirrel"
 	"github.com/google/go-cmp/cmp"
 	"github.com/stretchr/testify/assert"
+	"golang.org/x/exp/maps"
```
I guess this leads to all the go.mod changes? :(

alexzurbonsen replied:
Yes... it is just used to merge two maps. Could also write our own method for that?

sebathomas commented on lte/cloud/go/services/subscriberdb/storage/subscriber_state_test.go:
```diff
@@ -101,4 +105,82 @@ func TestSubscriberStorage(t *testing.T) {
+	})
+}
+
+func (ss *subscriberStorage) setTestGatewaySubscriberStateWithIncreasingTimestamp(t *testing.T, networkID string, gatewayID string, subscriberStates *GatewaySubscriberState, frozenTime int64) error {
```
I think the name of this function is based on an old version, don't see the increasing timestamp any more.

sebathomas commented on lte/cloud/go/services/subscriberdb/obsidian/handlers/handlers_test.go:
```diff
@@ -2316,3 +2358,14 @@ func newPolicy(id string) *policydbModels.PolicyRule {
+
+// newTestSubscriberStorage sets up the gateway_subscriber_states table, without
+// spinning up the entire subscriberdb Test Service, which could be done
+// alternatively.
```
```suggestion
// spinning up the entire subscriberdb Test Service.
```

sebathomas commented on lte/cloud/go/services/subscriberdb/storage/subscriber_state.go:
```diff
@@ -182,6 +195,58 @@ func (ss *subscriberStorage) DeleteSubscribersForGateway(networkID string, gatew
+		err = rows.Err()
+		if err != nil {
+			return nil, fmt.Errorf("GetSubscribersForIMSIs, SQL rows error: %w", err)
+		}
```
Idea: Maybe this could be extracted to a new method `convertRowsToMap` (or similar).

sebathomas commented on lte/cloud/go/services/subscriberdb/obsidian/handlers/handlers.go:
```diff
@@ -901,6 +918,20 @@ func mergeStates(s1 state_types.StatesByID, s2 state_types.StatesByID) state_typ
 	return s1
 }
 
+func getSubscribersForIMSIs(networkID string, imsis []string, subscriberStorage subscriberstorage.SubscriberStorage) (state_types.StatesByID, error) {
```
Maybe a better name:
```suggestion
func getSubscriberStatesForIMSIs(networkID string, imsis []string, subscriberStorage subscriberstorage.SubscriberStorage) (state_types.StatesByID, error) {
```

sebathomas reviewed (commented):
Looking good, just some minor points.
